### PR TITLE
[SPARK-35630][SQL] ExpandExec should not introduce unnecessary exchanges

### PR DESCRIPTION
**Pull request by tanelk**
### What changes were proposed in this pull request?

Improved `ExpandExec` so it would retain its child's outputPartitioning, when possible. 

### Why are the changes needed?

Currently `ExpandExecs` `outputPartitioning` is always `UnknownPartitioning(0)`. In some cases we do actually know the correct output partitioning. In those cases we could reduce the number of exchanges by using the correct partitioning.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

New UT

### CODE REVIEW

**Comment by tanelk** [comment]
@maropu , could you take a look at this?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/ExpandExec.scala [hunk 1]
```diff
@@ -42,9 +42,27 @@ case class ExpandExec(
...
+   * the output of the Expand.
+   * In general case the Expand can output data with arbitrary partitioning, so set it
+   * as UNKNOWN partitioning.
+   */
+  override def outputPartitioning: Partitioning = {
+    val stableAttrs = ExpressionSet(output.zipWithIndex.filter {
+      case (attr, i) => projections.forall(_(i).semanticEquals(attr))
+    }.map(_._1))
+
+    child.outputPartitioning match {
```
Comment (maropu) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/ExpandExec.scala:58]: How about the `PartitioningCollection` case?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/ExpandExec.scala [hunk 2]
```diff
@@ -42,9 +42,27 @@ case class ExpandExec(
...
+  /**
+   * The Expand is commonly introduced by the RewriteDistinctAggregates optimizer rule.
+   * In that case there can be several attributes that are kept as they are by the Expand.
+   * If the child's output is partitioned by those attributes, then so will be
+   * the output of the Expand.
+   * In general case the Expand can output data with arbitrary partitioning, so set it
+   * as UNKNOWN partitioning.
+   */
+  override def outputPartitioning: Partitioning = {
+    val stableAttrs = ExpressionSet(output.zipWithIndex.filter {
```
Comment (maropu) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/ExpandExec.scala:54]: Could you make this var name clearer? e.g., `passThroughtAttrs`?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/ExpandExec.scala [hunk 3]
```diff
@@ -42,9 +42,27 @@ case class ExpandExec(
   override lazy val metrics = Map(
     "numOutputRows" -> SQLMetrics.createMetric(sparkContext, "number of output rows"))
 
-  // The GroupExpressions can output data with arbitrary partitioning, so set it
-  // as UNKNOWN partitioning
-  override def outputPartitioning: Partitioning = UnknownPartitioning(0)
+  /**
+   * The Expand is commonly introduced by the RewriteDistinctAggregates optimizer rule.
```
Comment (maropu) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/ExpandExec.scala:46]: What does this comment mean? I think we have the three cases for `Expand`: distinct aggregates, grouping analytics, and window aggregates.

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala [hunk 4]
```diff
@@ -4003,6 +4003,56 @@ class SQLQuerySuite extends QueryTest with SharedSparkSession with AdaptiveSpark
...
   }
+
+  test("SPARK-35630: ExpandExec should not introduce unnecessary exchanges") {
+    withTable("test_table") {
+      spark.range(11)
+        .withColumn("group1", $"id" % 2)
+        .withColumn("group2", $"id" % 4)
+        .withColumn("a", $"id" % 3)
+        .withColumn("b", $"id" % 6)
+        .write.saveAsTable("test_table")
```
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala:4014]: nit: for better test performance, could you use a temporary view instead?

**Review by maropu — COMMENTED** [review]
cc: @viirya @cloud-fan

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala (hunk 4 above)
Comment (tanelk) [reply sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: I simplified it even a bit further

File: sql/core/src/main/scala/org/apache/spark/sql/execution/ExpandExec.scala (hunk 1 above)
Comment (tanelk) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/ExpandExec.scala]: Added that case, but I was not able to construct a test case for it

Comment (maropu) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/ExpandExec.scala]: we cannot use the shuffled join path for tests?  https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/joins/ShuffledJoin.scala#L48-L49

Comment (tanelk) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/ExpandExec.scala]: The `ProjectExec`, that was inserted after join, managed to simplify it to `HashPartitioning` for the cases I could think of.

Comment (maropu) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/ExpandExec.scala]: Ah, I see.

**Review by viirya — COMMENTED** [review]
The idea looks okay.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/ExpandExec.scala [hunk 5]
```diff
@@ -42,9 +42,33 @@ case class ExpandExec(
...
+
+    def getOutputPartitioning(partitioning: Partitioning): Partitioning = {
+      partitioning match {
+        case HashPartitioning(exprs, _) if exprs.forall(passThroughAttrs.contains) =>
+          partitioning
+        case RangePartitioning(ordering, _)
+          if ordering.map(_.child).forall(passThroughAttrs.contains) =>
+          partitioning
+        case PartitioningCollection(partitionings) =>
+          PartitioningCollection(partitionings.map(getOutputPartitioning))
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/ExpandExec.scala:65]: Do we have partitioning collection case for `Expand`?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/ExpandExec.scala [hunk 6]
```diff
@@ -42,9 +42,33 @@ case class ExpandExec(
   override lazy val metrics = Map(
     "numOutputRows" -> SQLMetrics.createMetric(sparkContext, "number of output rows"))
 
-  // The GroupExpressions can output data with arbitrary partitioning, so set it
-  // as UNKNOWN partitioning
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/ExpandExec.scala:46]: Shall we leverage `AliasAwareOutputPartitioning` and define the `outputExpressions` as the shared prefix of `projections`?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/ExpandExec.scala (hunk 5 above)
Comment (viirya) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/ExpandExec.scala:65]: I think we should. Just being curious on if there is an example case, maybe a test case too?

BTW, should we remove any unknown partition from `partitionings.map(getOutputPartitioning)`?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/ExpandExec.scala (hunk 6 above)
Comment (tanelk) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/ExpandExec.scala:46]: I tried it, seems to work fine. Lets see if all other tests pass too.

Comment (tanelk) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/ExpandExec.scala:46]: @cloud-fan Could you take another look at this

**Pull request closed without merging**